libraries/lib-music-information-retrieval/GetMeterUsingTatumQuantizationFit.cpp

```diff
@@ -0,0 +1,476 @@
+   // train of frequency `numTatums / odf.size()`. We take the position of the
+   // first peak to be the lag.
+   const auto odfSamplesPerDiv = 1. * odf.size() / numTatums;
+   auto max = std::numeric_limits<float>::lowest();
+   auto lag = 0;
+   std::vector<int> tatumIndices(numTatums);
+   std::iota(tatumIndices.begin(), tatumIndices.end(), 0);
+   while (true)
+   {
+      const auto val = std::accumulate(
+         tatumIndices.begin(), tatumIndices.end(), 0.f, [&](float sum, int i) {
+            const int j = std::round(i * odfSamplesPerDiv) + lag;
+            return sum + (j < odf.size() ? odf[j] : 0.f);
+         });
+      if (val < max)
+         break;
+      max = val;
+      ++lag;
+   }
+   return lag - 1;
+}
+
+// This is the fundament of the algorithm. It gives a weighted average of the
+// normalized distance between ODF peaks and the closest tatum. The weights are
+// the ODF values at the peaks. The distance is normalized by the tatum
+// duration, so that the final value ranges between 0 and 1.
+// This can be seen as a performance accuracy measure. The higher the tempo, the
+// more accurate the performer has to be for this distance to stay low. In
+// general, this is desired. We could nevertheless consider introducing a slight
+// tolerance which increases with the tatum rate, as very fast rhythmic patterns
+// recorded without post-editing still tend to achieve lower scores.
+double GetQuantizationDistance(
+   const std::vector<int>& peakIndices, const std::vector<float>& peakValues,
+   size_t size, int numDivisions, int lag)
+{
+   std::vector<double> peakDistances(peakIndices.size());
+   const auto peakSum = std::accumulate(
+      peakValues.begin(), peakValues.end(), 0., std::plus<double> {});
+   const auto odfSamplesPerDiv = 1. * size / numDivisions;
+   std::transform(
+      peakIndices.begin(), peakIndices.end(), peakDistances.begin(),
+      [&](int peakIndex) {
+         const auto shiftedIndex = peakIndex - lag;
+         const auto closestDiv = std::round(shiftedIndex / odfSamplesPerDiv);
+         // Normalized distance between 0 and 1:
+         const auto distance = 2 *
+                               (shiftedIndex - closestDiv * odfSamplesPerDiv) /
+                               odfSamplesPerDiv;
+         return distance < 0 ? -distance : distance;
+      });
+   // Calculate the score as the sum of the distances weighted by
+   // the odf values:
+   const auto weightedAverage =
+      std::inner_product(
+         peakDistances.begin(), peakDistances.end(), peakValues.begin(), 0.,
+         [](double a, double b) { return a + b; },
+         [&](double distance, float peakValue) {
+            return distance * peakValue;
+         }) /
+      peakSum;
+   return weightedAverage;
+}
+
+// Look at the amplitude across all scores. If our ODF is that of a musical
+// rhythm, then the contrast between good and bad quantizations should make a
+// big difference.
+double GetExperimentContrastMeasure(
```

Comment:
I just removed the contrast measure in a follow-up commit, improving performance of the classifier and simplifying the procedure. Please ignore this part while reviewing.